VAOS Calendar: Show message for max selections and add screen reader tags for validation error

narin opened the pull request:
## Description
On the request calendar, if the user has 3 date/time combinations selected and tries to select a 4th, display a validation error.

## Testing done


## Screenshots
![image](https://user-images.githubusercontent.com/786704/75569982-0ac45300-5a0b-11ea-8ff4-6f8e4c491c39.png)


## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

narin committed:
Add disabled style to other checkboxes if max selections reached.

Remove currentlySelectedRow and isValid from CalendarWidget since they are unused.

narin committed:
Update validation copy.  Add sr-only text to calendar option labels

narin commented on src/applications/vaos/actions/newAppointment.js:
```diff
@@ -523,17 +523,12 @@ export function getAppointmentSlots(startDate, endDate) {
     calendarData: {
       currentlySelectedDate,
       selectedDates,
-      currentRowIndex,
```
Remove `currentRowIndex` from `calendarData` in redux state since it is no longer used.  Also removed references to it throughout this PR.

narin commented on src/applications/vaos/components/calendar/CalendarCheckboxOption.jsx:
```diff
@@ -1,29 +1,47 @@
+      'vads-u-color--primary': !disabled,
+      'vads-u-color--gray-lighter': disabled,
+    },
+  );
```
Use CSS styles to make the checkboxes appear disabled.  Had to do this instead of adding the `disabled` attribute because we still want to be able to detect click events on the disabled checkboxes.

1Copenut replied:
I saw that when I was reviewing this morning. It validates to have `disabled` on the `<div>` even though we're _not exactly_ disabling it. I've done some fiddling with the CSS to add more contrast to the "disabled" text and make sure the not-allowed cursor extends to the entire label not just the input hit space. Attaching a screen shot with notes below. /cc @rtwell 

---
![css-notes-calendar-widget](https://user-images.githubusercontent.com/934879/76103671-acade780-5f97-11ea-8c0b-071bc398ccb4.jpg)

narin commented on src/applications/vaos/components/calendar/CalendarWidget.jsx:
```diff
@@ -105,85 +103,27 @@ export default class CalendarWidget extends Component {
-    }
-
-    return isValid;
-  };
```
Removed this since we are performing validations on the page itself, rather than within the `CalendarWidget`

narin committed:
Scroll and focus for max selections error

narin committed:
Use color util

narin committed:
Fix color util

narin committed:
Update Calendar option tests

narin committed:
Update CalendarOptions test

narin committed:
Add tests for DateTimeRequestPage

narin committed:
Add tests for CalendarRow

narin committed:
Add tests for CalendarCell

narin committed:
Add maxSelections test

narin committed:
Add tests to CalendarCell

narin commented:
diff-coverage is mainly being brought down by `CalendarWidget.jsx`. We have a separate ticket (https://github.com/department-of-veterans-affairs/va.gov-team/issues/4776) to improve unit tests for this file.

jbalboni commented:
This is maybe more of a @rtwell or @1Copenut question, but this UI seems really confusing. When a checkbox click does nothing, that's typically an error, unless the box is disabled. 

Also, showing a validation error when there's not anything for a user to correct is contrary to the rest of our forms and would I think be confusing to anyone trying to move to the next page. At this point a user has gone through several pages where this red error UI means they need to fix something before continuing, but now we expect them to know there's nothing to fix and they should continue. If I encountered that, I'd think it was a bug in the form.

jbalboni commented on src/applications/vaos/components/calendar/CalendarWidget.jsx:
```diff
@@ -105,85 +103,27 @@ export default class CalendarWidget extends Component {
+    // If an option selection is required, don't add this date to selectedDates
+    // array until an option is selected as well
+    if (!isInSelectedArray && !this.props.additionalOptions?.required) {
+      selectedDates.push({ date });
```
Nitpicky and not new, but we shouldn't modify this array because it's in Redux and we should treat that data as immutable. You could use `concat` instead.

narin replied:
`selectedDates` is declared above using the spread operator.  This creates a new copy of the array, correct? So we wouldn't be mutating the data:

```
const selectedDates = this.props.selectedDates
      ? [...this.props.selectedDates]
      : [];
```

jbalboni replied:
Yep, you're right. The weird github diff gap threw me off.

rtwell commented:
> This is maybe more of a @rtwell or @1Copenut question, but this UI seems really confusing. When a checkbox click does nothing, that's typically an error, unless the box is disabled.

isn't the error state (previewed above) being triggered when the 4th checkbox is clicked? or am i misunderstanding the interaction?

jbalboni commented:
@rtwell At the moment it looks like we're not allowing you to check four boxes. When you attempt to check the fourth, the validation error shows up and the checkbox stays the same.

narin commented:
> isn't the error state (previewed above) being triggered when the 4th checkbox is clicked? or am i misunderstanding the interaction?

@rtwell Attempting to click on a 4th checkbox displays the error, but we don't actually let them have 4 checked (all checkboxes are disabled if 3 selected already).  So even though there is an error displaying, the user still only has 3 checked, which is technically valid and the user can continue.  However since there is an error being displayed, this could be confusing.

rtwell commented:
ah, I understand, thank you. Let's allow the user to select a 4th box, trigger the error when they do check the 4th box, and prevent them from advancing before un-checking one of the 4 boxes. Does that make sense/line up with your expectations?

narin commented:
@rtwell I think that works but I'm wondering if @1Copenut has any thoughts?

1Copenut commented on src/applications/vaos/components/calendar/CalendarCheckboxOption.jsx:
```diff
@@ -1,29 +1,47 @@
+      <label className={labelClasses} htmlFor={`checkbox-${id}`}>
+        <span aria-hidden="true">{label}</span>
+        <span className="vads-u-visibility--screen-reader">
+          {label} option selected
```
This label was confusing when I listened to it on a screen reader. What I heard was: "AM option selected, checkbox not checked" or "AM option selected, checkbox checked". I think it'd be worth reaching out to a content specialist for the right language. Maybe something like "AM appointment" which would then read "AM appointment, checkbox (un)checked"

/cc @peggygannon

1Copenut commented on src/applications/vaos/containers/DateTimeRequestPage.jsx:
```diff
@@ -42,19 +47,18 @@ export class DateTimeRequestPage extends React.Component {
+      (this.isMaxSelectionsError() ||
+        (!prevState.submitted && this.state.submitted))
     ) {
       scrollAndFocus('.usa-input-error-message');
```
I'll add my screenshot here. Let's remove the yellow focus halo from the `<span class="usa-input-error-message">` because it's not a typical focusable element. I'd also like to add `role="alert"` to that span, so it announces the error message assertively to all screen readers. VoiceOver missed it once because it was still reading out the input instructions.

---
![Screen Shot 2020-03-06 at 10 22 45 AM](https://user-images.githubusercontent.com/934879/76104114-6442f980-5f98-11ea-8512-cb924f136359.png)

1Copenut reviewed (changes requested):
@narin && @lenaecb I reviewed the error handling with Safari + VoiceOver this morning locally. Have 3 changes and tagged Ryan and Peggy for guidance on 2 of them respectively. I'll retest when Narin's ready on VO, then in Staging on NVDA and JAWS as soon as code is promoted.

narin committed:
Merge branch 'master' into 5992-cal-request-validation

narin committed:
Allow user to select more than 3 options and show validation error.  Update checkbox styling.  Remove outline from alert and add role="alert"

narin committed:
Fix test

narin commented:
Hey @1Copenut Updated styling for checkbox and containers, updated SR text, removed outline from validation text, and added `role="alert"` as per your recommendations.

narin committed:
Fix test

1Copenut reviewed (changes requested):
@narin The updated color and styles for the non-selected radios looks good. I noticed one issue where I could go ahead and click the fourth checkbox. The error fires, but the checkbox turns to our selected style. Screenshots attached.

---
![Screen Shot 2020-03-17 at 2 08 16 PM](https://user-images.githubusercontent.com/934879/76892774-35126f00-6859-11ea-8368-62592e713ae9.png)

---
![Screen Shot 2020-03-17 at 2 08 37 PM](https://user-images.githubusercontent.com/934879/76892792-3b085000-6859-11ea-941d-5410d2331980.png)

narin commented:
@1Copenut Thanks for the review.  Being able to check a 4th checkbox is expected and is what throws that validation error, as per the discussion we had a few days ago in Slack. Are you suggesting the 4th checkbox styling be different?

1Copenut reviewed (approved):
After further discussion with @narin I understand the intent. Things are functioning correctly. Moving to approve with a quick note to @rtwell that we could consider revisiting this user flow post-launch. Being able to check a fourth box, see the error, then have to remove the fourth checked box before I could proceed could be slightly confusing.

rtwell commented:
agreed, @1Copenut — dismiss the 4th check and then continue is clearer.

jbalboni commented on src/applications/vaos/tests/containers/DateTimeRequestPage.unit.spec.jsx:
```diff
@@ -47,13 +47,14 @@ describe('VAOS <DateTimeRequestPage>', () => {
     const onCalendarChange = sinon.spy();
     const routeToNextAppointmentPage = sinon.spy();
+    const data = {
+      calendarData: { currentlySelectedDate: '2020-12-20', error: 'test' },
```
Could we set this date (and the ones below) to today's date, plus a year (or some other interval)? That seems less likely to maybe cause a time-based unit test failure down the road.

jbalboni reviewed (approved):
One minor comment, LGTM otherwise. Worked fine locally, too.

narin committed:
Update test to use moment

narin committed:
Merge branch 'master' into 5992-cal-request-validation
